main.cpp

```diff
@@ -0,0 +1,205 @@
+
+#include <algorithm>
+#include <execution>
+#include <iostream>
+#include <llvm/IR/DerivedTypes.h>
+#include <memory>
+#include <pstl/glue_execution_defs.h>
+#include <string>
+#include <system_error>
+
+#include <llvm/ADT/SmallVector.h>
+#include <llvm/ADT/StringExtras.h>
+#include <llvm/IR/Argument.h>
+#include <llvm/IR/DerivedTypes.h>
+#include <llvm/IR/Function.h>
+#include <llvm/IR/GlobalValue.h>
+#include <llvm/IR/Instructions.h>
+#include <llvm/IR/Module.h>
+#include <llvm/Support/Casting.h>
+#include <llvm/Support/CommandLine.h>
+#include <llvm/Support/SourceMgr.h>
+#include <llvm/Support/raw_ostream.h>
+
+#include <llvm/ADT/StringRef.h>
+#include <llvm/Bitcode/BitcodeWriter.h>
+#include <llvm/IRReader/IRReader.h>
+#include <llvm/Transforms/Utils/Cloning.h>
+#include <llvm/Transforms/Utils/ValueMapper.h>
+
+static llvm::LLVMContext context;
+
+enum FunctionAction { MakePublic, Skip, MakePrivate, MakeRefernce };
+
+static void copyComdat(llvm::GlobalObject *Dst, const llvm::GlobalObject *Src) {
+  const auto *SC = Src->getComdat();
+  if (!SC)
+    return;
+  auto *DC = Dst->getParent()->getOrInsertComdat(SC->getName());
+  DC->setSelectionKind(SC->getSelectionKind());
+  Dst->setComdat(DC);
+}
+
+void moveFunctions(const llvm::StringRef outputName, const llvm::Module *module,
+                   FunctionAction(CheckFunction)(const llvm::Module &,
+                                                 const llvm::Function &)) {
+  llvm::Module result(llvm::StringRef("output"), context);
+
+  result.setDataLayout(module->getDataLayout());
+  result.setTargetTriple(module->getTargetTriple());
+  result.setModuleInlineAsm(module->getModuleInlineAsm());
+
+  llvm::ValueToValueMapTy VMap;
+
+  for (const auto &alias : module->aliases()) {
+    if (!alias.getValueType()->isFunctionTy()) {
+      auto *GA = llvm::GlobalAlias::create(
+          alias.getValueType(), alias.getType()->getPointerAddressSpace(),
+          alias.getLinkage(), alias.getName(), &result);
+      GA->copyAttributesFrom(&alias);
+      VMap[&alias] = GA;
+    }
+  }
+
+  for (const auto &global : module->globals()) {
+    std::cout << global.getName().str() << std::endl;
+
+    auto GV = new llvm::GlobalVariable(
+        result, global.getValueType(), global.isConstant(), global.getLinkage(),
+        nullptr, global.getName(), nullptr, global.getThreadLocalMode(),
+        global.getType()->getAddressSpace());
+
+    // TODO: Find a way to understand how CloneModule copies these things
```

Comment:
I suspect that `CloneModule` is meant for use in the earlier stages of compilation, prior to the target-specific passes which typically set the required alignment, and so does attempt to preserve alignment.